---
title: "SNV Calling and VCF files"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
---

# Introduction

In this session we:-

- Demonstrate how to call germline variants for a bam file containing aligned reads
- Explain what a *vcf* file is
- Loading a *vcf* file into IGV
- Interrogating vcf files using Bioconductor
- Intersect sets of calls

## Calling SNVs

We will use the [freebayes](https://github.com/ekg/freebayes#readme) genotype caller for convenience. The main reasons for choosing this were:-

- freely-available with an open-source licence
- easy to run and minimal command-line parameters to specify
- it is Bayesian; which we like
- gives reasonable results

`freebayes` is a command-line tool that you will need to run from the terminal. As with all command-line tools, we first need to make sure that we are located in the correct directory. Remember that we can *change directory* using the unix command `cd`.


`freebayes` has already been in the ***CRUK Docker***. However, we can make sure that `freebayes` is available by typing `freebayes` at the terminal. This will give a basic description, a version number, and information about how to cite the package (very important if you end-up using the tool to generate results for your paper). Adding the *parameter* `-h` will display more-detailed help about the tool and some examples of how to use it.


```{}
cd /home/participant/Course_Materials/Day2/
freebayes
freebayes -h
```

We are going to use `freebayes` to call SNVs on some *1000 genomes* samples. In order to make the SNV-calling run in a reasonable time, we are only considering reads aligned to chromosome 20 in this analysis. In the [Appendix](#appendix) you will see the commands used to create the bam files.

The minimal requirements to run `freebayes` (and why it is appealing for this practical!) are a reference genome and a `.bam` file. The `-f` parameter is used to specify the location of a reference genome in `.fasta` format. 


***Please don't run this next command*** 

```{}
freebayes -f /reference_data/Homo_sapiens_assembly19.fasta /data/hapmap/NA12878.chr20.bam
```

If you did run that command, you would quickly see that the screen gets filled with lots of text. These are the calls that `freebayes` is making being printed to the screen (the standard output for some unix commands). If you find yourself in this situation, a swift press of `CTRL + C` should stop the tool from running.

What we need to do is *direct the output* to a file. We can call the output file anything we like, but it is advisable to make the name relatable to the name of the input. If we are in the situation of calling genotypes on many samples, with many different callers, then we want to be able to identify the processing used on each sample.

Although it is not mandatory, we give the output file the extension `.vcf`. The `.vcf` format is a commonly-adopted standard for variant calls, which we will look into detail now.

We also restrict the analysis to a particular region to speed things up
```{}
freebayes -f /reference_data/Homo_sapiens_assembly19.fasta --region 20:500000-800000 /data/hapmap/NA12878.chr20.bam > NA12878.chr20.subset.freebayes.vcf
```

## Understanding the vcf format

The vcf format was initially developed by the [1000 Genomes Project](http://www.1000genomes.org/wiki/Analysis/vcf4.0), and ownership has been subsequently transferred to [Global Alliance for Genomics and Health Data Working group file format team](http://ga4gh.org/#/fileformats-team). The format can be used to represent information about all kinds of genomic variation. In this session we will just consider SNVs.

We don't require any specialised software to look at the contents of a vcf file. They *can* be opened in a bog-standard text editor.For now, navigate to the folder and right-click on `NA12878.chr20.subset.freebayes.vcf`. Select the option to open with `gedit` (a common text editor for Ubuntu).


In a similar vein to the `.bam` and `.sam` files we saw earlier, the `.vcf` files contains many lines of header information.
```{r echo=FALSE,comment=NA}
cat(system("head NA12878.chr20.subset.freebayes.vcf",intern=TRUE),sep="\n")

```


After many more lines of information, we finally get to the details of the actual calls themsevles. This part of the file is tab-delimited; with 10 columns for every call. The [vcf specification page](http://www.1000genomes.org/wiki/Analysis/vcf4.0) gives details of what should be contained in each column

- CHROM
- POS
- ID
- REF
- ALT
- QUAL
- FILTER
- INFO
- FORMAT
- NA12878

Shown here is the information about the first five calls

```{r echo=FALSE,comment=NA}
cat(system("grep -v '##' NA12878.chr20.subset.freebayes.vcf | head -n5",intern = TRUE),sep = "\n")
```


The first seven columns should look consistent across different genotype callers. In this exercise we have not annotated against known variants, or applied any filtering, so the `ID` and `FILTER` columns are blank. In some `.vcf` files these columns might be populated with dbSNP IDs or flags such as `PASS` / `FAIL` respectively.

The contents of the `INFO` and `FORMAT` columns will depend on what variant caller has been used. The `INFO` column contains metrics and other information related to ***each variant call*** as a set of `KEY=VALUE` pairs. Each pair is separated by a `;` character. 

The `INFO` for the first variant reads as:-

`AB=0.454545;ABP=3.20771;AC=1;AF=0.5;AN=2;AO=5;CIGAR=1X;DP=11;DPB=11;DPRA=0;EPP=3.44459;EPPR=4.45795;GTI=0;.......`

which we can interpret as:- 

```{r echo=FALSE}
library(stringr)
info <- "AB=0.454545;ABP=3.20771;AC=1;AF=0.5;AN=2;AO=5;CIGAR=1X;DP=11;DPB=11;DPRA=0;EPP=3.44459;EPPR=4.45795;GTI=0;LEN=1;MEANALT=1;MQM=60;MQMR=60;NS=1;NUMALT=1;ODDS=21.8887;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=188;QR=245;RO=6;RPL=1;RPP=6.91895;RPPR=3.0103;RPR=4;RUN=1;SAF=2;SAP=3.44459;SAR=3;SRF=5;SRP=8.80089;SRR=1;TYPE=snp;technology.ILLUMINA=1"

tmp <- sapply(str_split(info, ";")[[1]],function(x) str_split(x, "="))
df <- t(do.call(cbind, tmp))
colnames(df) <- c("Key","Value")
knitr::kable(df[1:10,])

```


The meaning of each `KEY`  is given in the header for the `.vcf` file.

```{r echo=FALSE}
cat(system("grep 'ID=AB,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=ABP,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AC,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AF,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AN,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AO,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=CIGAR,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DP,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DPB,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DPRA,' NA12878.chr20.subset.freebayes.vcf | head -n1",intern=TRUE))

```



The final column in the file describes the calls for the sample `NA12878`, which was the only sample that we called variants on. In the sample column (`NA12878`) for the first variant we see the entry 

(`0/1:11:11,5:6:245:5:188:-13.9693,0,-19.1141`). 

These are values separated by a `:` character and they are interpreted in the same order as dictated by the `FORMAT` column which is `GT:DP:DPR:RO:QR:AO:QA:GL`

```{r echo=FALSE}
gt <- "0/1:11:11,5:6:245:5:188:-13.9693,0,-19.1141"
ky <- "GT:DP:DPR:RO:QR:AO:QA:GL"
values <- unlist(str_split(gt, ":"))
keys <- unlist(str_split(ky, ":"))
desc <- c("Genotype","Read Depth","Number of observation for each allele","Reference allele observation count","Sum of quality of the reference observations","Alternate allele observation count","Sum of quality of the alternate observations","Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy")
df <- data.frame(keys,values,desc)
colnames(df) <- c("Key", "Value","Description")
knitr::kable(df)
```


So for this particular variant, in the sample `NA12878` there is a genotype of `0\1` (heterozygous) and a depth of 11 etc.

To understand the vcf format better, we can use another mode of operation of `freebayes` which is to call genotypes on multiple samples simultaneously. We can also specify an exact region of the genome to call genotypes to speed-up the processing.

```{}
freebayes -f /reference_data/Homo_sapiens_assembly19.fasta --region 20:500000-800000 /data/hapmap/NA12878.chr20.bam /data/hapmap/NA12873.chr20.bam /data/hapmap/NA12874.chr20.bam > combined.chr20.subset.freebayes.vcf

```

As before, we can look at the first five calls.

```{r echo=FALSE,comment=NA}
cat(system("grep -v '##' combined.chr20.subset.freebayes.vcf | head -n5",intern = TRUE),sep = "\n")
```

You should notice that we now have columns `NA12878`, `NA12873` and `NA12874` as we have called variants in three samples. 

One easy way of being able to visualise the calls is to use the IGV browser. Before we do this however, we can do some extra processing to make it easier to process the files. This series of commands will compress and index the vcf files (similar to how bam files are indexed to produce a `.bai` file). IGV would probably cope fine with reading such relatively-small files, but it is good to get into the habit of processing our files in this manner.

```{}

bgzip -c combined.chr20.subset.freebayes.vcf > combined.chr20.subset.freebayes.vcf.gz
tabix -p  vcf combined.chr20.subset.freebayes.vcf.gz

bgzip -c NA12878.chr20.subset.freebayes.vcf > NA12878.chr20.subset.freebayes.vcf.gz
tabix -p vcf NA12878.chr20.subset.freebayes.vcf.gz

```

After running these commands, you should see that the files `combined.chr20.subset.freebayes.vcf.gz`, `combined.chr20.subset.freebayes.vcf.gz.tbi`, `NA12878.chr20.subset.freebayes.vcf.gz` and `NA12878.chr20.subset.freebayes.vcf.gz.tbi` have been created in your working directory.

These files can be viewing in IGV and as usual we can zoom-in and scroll along the genome. Each `.vcf` introduces two tracks into the IGV data panel; the first gives information about the variant, and the second is for sample-specific information.

******
******
******

### Exercise

- Load IGV
- Select the files `NA12878.chr20.subset.freebayes.vcf.gz` and `combined.chr20.subset.freebayes.vcf.gz` from the `File -> Open` menu
- Navigate to `chr20:500,900-505,630`
- Verify that the same information from the `.vcf` is shown in IGV
- What do the light and dark blue rectangles in the sample-specific tracks refer to?
- What SNPs are shared between the different samples?
    + are any unique to a particular sample?
- Load the corresponding `bam` files to complement our understanding of the calls

    
******
******
******

To dig-into these files further, we will use tools within Bioconductor. Our goal will be to find calls that are in common between different samples, and what SNPs occur within coding regions. For completeness, we note that other command-line tools can be used to perform these operations (the [bedtools](http://bedtools.readthedocs.io/en/latest/) suite for example). However, we should be reasonably comfortable with R by now and the package we use interacts nicely with the `GenomicRanges` concepts we discussed yesterday.
