---
title: "Working with VCF files"
author: "Mark Dunning"
date: "24 May 2016"
output: html_document
---

# Introduction

In this session we:-

- Demonstrate how to call germline variants for a bam file containing aligned reads
- Explain what a *vcf* file is
- Loading a *vcf* file into IGV
- Interrogating vcf files using Bioconductor
- Intersect sets of calls

## Calling SNVs

We will use the [freebayes](https://github.com/ekg/freebayes#readme) genotype caller for convenience. The main reasons for choosing this were:-

- freely-available with an open-source licence
- easy to run and minimal command-line parameters to specify
- gives reasonable results

`freebayes` is a command-line tool that you will need to run from the terminal. As with all command-line tools, we first need to make sure that we are located in the correct directory. Remember that we can *change directory* using the unix command `cd`.

```{}
cd ~/Course_Materials/Day2/
```

`freebayes` has already been installed for you on your desktop machine (or in the *docker* image, if you are using that). However, we can make sure that `freebayes` is available by typing `freebayes` at the terminal. This will give a basic description, a version number, and information about how to cite the package (very important if you end-up using the tool to generate results for your paper). Adding the *parameter* `-h` will display more-detailed help about the tool and some examples of how to use it.

```{}
freebayes
freebayes -h
```

We are going to use `freebayes` to call SNVs on some *hapmap* samples. In order to make the SNV-calling run in a reasonable time, we are only considering reads aligned to chromosome 20 in this analysis. In the [Appendix](#appendix) you will see the commands used to create the bam files.

The minimal requirements to run `freebayes` (and why it is appealing for this practical!) are a reference genome and a `.bam` file. The `-f` parameter is used to specify the location of a reference genome in `.fasta` format. 

***Please don't run this next command*** 

```{}
freebayes -f ../ref_data/human_g1k_v37.fasta data/hapmap/NA12878.chr20.bam
```

If you did run that command, you would quickly see that the screen gets filled with lots of text. These are the calls that `freebayes` is making being printed to the screen (the standard output for some unix commands). If you find yourself in this situation, a swift press of `CTRL + C` should stop the tool from running.

What we need to do is *direct the output* to a file. We can call the output file anything we like, but it is advisable to make the name relatable to the name of the input. If we are in the situation of calling genotypes on many samples, with many different callers, then we want to be able to identify the processing used on each sample.

Although it is not mandatory, we give the output file the extension `.vcf`. The `.vcf` format is a commonly-adopted standard for variant calls, which we will look into detail now.

```{}
freebayes -f ../ref_data/human_g1k_v37.fasta data/hapmap/NA12878.chr20.bam > NA12878.chr20.freebayes.vcf
```

## Understanding the vcf format

The vcf format was initially developed by the [1000 Genomes Project](http://www.1000genomes.org/wiki/Analysis/vcf4.0), and ownership has been subsequently transferred to [Global Alliance for Genomics and Health Data Working group file format team](http://ga4gh.org/#/fileformats-team).

We don't require any specialised software to look at the contents of a vcf file. They *can* be opened in a bog-standard text editor. However, there are several tools that will make our lives easier! For now, navigate to the folder and right-click on `NA12878.chr20.freebayes.vcf`. Select the option to open with `gedit` (a common text editor for Ubuntu).


In a similar vein to the bam and sam files we saw earlier, the `.vcf` files contains many lines of header information.
```{r echo=FALSE,comment=NA}
cat(system("head NA12878.chr20.freebayes.vcf",intern=TRUE),sep="\n")

```


After many more lines of information, we finally get to the details of the actual calls themsevles. This part of the file is tab-delimited; with 10 columns for every call. The [vcf specification page](http://www.1000genomes.org/wiki/Analysis/vcf4.0) gives details of what should be contained in each column

- CHROM
- POS
- ID
- REF
- ALT
- QUAL
- FILTER
- INFO
- FORMAT
- NA12878

Shown here is the information about the first five calls

```{r echo=FALSE,comment=NA}
cat(system("grep -v '##' NA12878.chr20.freebayes.vcf | head -n5",intern = TRUE),sep = "\n")
```


The first seven columns should look consistent across different genotype callers. In this exercise we have not annotated against known variants, or applied any filtering, so the `ID` and `FILTER` columns are blank. In some `.vcf` files these might be populated with dbSNP IDs or flags such as `PASS` / `FAIL`.

The contents of the `INFO` and `FORMAT` columns will depend on what variant caller has been used. The `INFO` column contains metrics and other information related to each variant call as a set of `KEY` pairs. Each pair is separated by a `;` character. 

The `INFO` for the first variant reads as:-

`AB=0.454545;ABP=3.20771;AC=1;AF=0.5;AN=2;AO=5;CIGAR=1X;DP=11;DPB=11;DPRA=0;EPP=3.44459;EPPR=4.45795;GTI=0;.......`

which we can interpret as:- 

```{r echo=FALSE}
library(stringr)
info <- "AB=0.454545;ABP=3.20771;AC=1;AF=0.5;AN=2;AO=5;CIGAR=1X;DP=11;DPB=11;DPRA=0;EPP=3.44459;EPPR=4.45795;GTI=0;LEN=1;MEANALT=1;MQM=60;MQMR=60;NS=1;NUMALT=1;ODDS=21.8887;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=188;QR=245;RO=6;RPL=1;RPP=6.91895;RPPR=3.0103;RPR=4;RUN=1;SAF=2;SAP=3.44459;SAR=3;SRF=5;SRP=8.80089;SRR=1;TYPE=snp;technology.ILLUMINA=1"

tmp <- sapply(str_split(info, ";")[[1]],function(x) str_split(x, "="))
df <- t(do.call(cbind, tmp))
colnames(df) <- c("Key","Value")
knitr::kable(df[1:10,])

```


The meaning of each `KEY`  is given in the header for the `.vcf` file.

```{r echo=FALSE}
cat(system("grep 'ID=AB,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=ABP,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AC,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AF,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AN,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=AO,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=CIGAR,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DP,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DPB,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))
cat(system("grep 'ID=DPRA,' NA12878.chr20.freebayes.vcf | head -n1",intern=TRUE))

```



The final column in the file describes the calls for the sample `NA12878`, which was the only sample that we called variants on. In the sample column (`NA12878`) for the first variant we see the entry 

(`0/1:11:11,5:6:245:5:188:-13.9693,0,-19.1141`). 

These are values separated by a `:` character and they are interpreted in the same order as dictated by the `FORMAT` column. 

```{r echo=FALSE}
gt <- "0/1:11:11,5:6:245:5:188:-13.9693,0,-19.1141"
ky <- "GT:DP:DPR:RO:QR:AO:QA:GL"
values <- unlist(str_split(gt, ":"))
keys <- unlist(str_split(ky, ":"))
desc <- c("Genotype","Read Depth","Number of observation for each allele","Reference allele observation count","Sum of quality of the reference observations","Alternate allele observation count","Sum of quality of the alternate observations","Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy")
df <- data.frame(keys,values,desc)
colnames(df) <- c("Key", "Value","Description")
knitr::kable(df)
```


So for this particular variant, in the sample `NA12878` there is a genotype of `0\1` (heterozygous) and a depth of 11.

To understand the vcf format better, we can use another mode of operation of `freebayes` which is to call genotypes on multiple samples simultaneously. We can also specify an exact region of the genome to call genotypes to speed-up the processing.

```{}
freebayes -f ../ref_data/human_g1k_v37.fasta --region 20:500000-800000 data/hapmap/NA12878.chr20.bam data/hapmap/NA12873.chr20.bam data/hapmap/NA12874.chr20.bam > combined.chr20.subset.freebayes.vcf

```

```{r echo=FALSE}
cat(system("grep -v '##' combined.chr20.subset.freebayes.vcf | head -n5",intern = TRUE),sep = "\n")
```



One easy way of being able to visualise the calls is to use the IGV browser. Before we do this however, we can do some extra processing to make it easier to process the files. This series of commands will compress and index the vcf files (similar to how bam files are indexed to produce a `.bai` file). IGV would probably cope fine with reading such relatively-small files, but it is good to get into the habit of processing our files in this manner.

```{}

bgzip -c combined.chr20.subset.freebayes.vcf > combined.chr20.subset.freebayes.vcf.gz
tabix -p  vcf combined.chr20.subset.freebayes.vcf.gz

bgzip -c NA12878.chr20.freebayes.vcf > NA12878.chr20.freebayes.vcf.gz
tabix -p vcf NA12878.chr20.freebayes.vcf.gz

```

After running these commands, you should see that the files `combined.chr20.subset.freebayes.vcf.gz`, `combined.chr20.subset.freebayes.vcf.gz.tbi`, `NA12878.chr20.freebayes.vcf.gz` and `NA12878.chr20.freebayes.vcf.gz.tbi` have been created in your working directory.

******

### Exercise

- Load IGV
- Select the files `NA12878.chr20.freebayes.vcf.gz` and `combined.chr20.subset.freebayes.vcf.gz` from the `File -> Open` menu
- Navigate to `chr20:60000-70000`
- Verify that the same information from the `.vcf` is shown in IGV
- What do the difference colours correspond to?
- What SNPs are shared between the different samples?
    + are any unique to a particular sample?
    
******

To dig-into these files further, we will use tools within Bioconductor. Our goal will be to find calls that are in common between the two different calling methods, and what SNPs occur within coding regions. For completeness, we note that other command-line tools can be used to perform these operations (the [bedtools](http://bedtools.readthedocs.io/en/latest/) suite for example). However, we should be reasonably comfortable with R by now and the package we use interacts nicely with the `GenomicRanges` concepts we discussed yesterday.

## Importing into Bioconductor

### Multi-sample example

The `VariantAnnotation` package allows `.vcf` files to be imported. The `readVcf` function can be used and requires the file of a vcf file. You also need to specify a genome name.

```{r message=FALSE}
library(VariantAnnotation)
hapmap.vcf <- readVcf("combined.chr20.subset.freebayes.vcf","hg19")

```

As we have seen already with other Bioconductor objects, typing the name of the object will print a summary to the screen. In the case of a "`CollapsedVCF`" object this is very detailed
```{r}
hapmap.vcf
```

The "header" information can be extracted using `header`. This will contain the definitions of all the per-genotype (`INFO`) and per-sample information stored in the file

```{r eval=FALSE}
header(hapmap.vcf)
info(header(hapmap.vcf))
geno(header(hapmap.vcf))
```

```{r echo=FALSE}
info(header(hapmap.vcf))[1:3,]
geno(header(hapmap.vcf))[1:3,]
```

## `INFO`

As described above, the `INFO` column in a `.vcf` file gives per-variant metadata as a series of key-value pairs. We can interrogate these data using the `info` function, which returns a data-frame-like object. The name ofe each row is derived from the 

```{r eval=FALSE}
info(hapmap.vcf)
```

```{r echo=FALSE}
info(hapmap.vcf)[1:3,1:5]
```

`NS` represents the number of samples with data. We would expect this to be 3 in all cases. However, as we are dealing with low-coverage data there are some sites data could not be observed for all samples

```{r}
summary(info(hapmap.vcf)$NS)
```

```{r eval=FALSE}
infoMat <- info(hapmap.vcf)
infoMat[info(hapmap.vcf)$NS==1,]
```

```{r echo=FALSE}
infoMat <- info(hapmap.vcf)
infoMat[info(hapmap.vcf)$NS==1,1:5]
```

`DP` gives the depth observed at each position, which is given to us in the form of a vector. By looking at the distribtion of the depths, we see some variants with extremely low coverage. Later-on, we will see how we can develop filters to remove unreliable calls from our dataset. Variants with low depth would seem to be a good candidate for exclusion.

```{r}
hist(infoMat$DP)
table(infoMat$DP)
```

The columns we have looked at, `DP` and `NS` have both returned vectors when we extract them from the data frame. However, this is not the case for all columns in the data frame.

The variable `TYPE` indicates whether each variant is an snp, insertion, deletion or something else. However, the format of this column is somewhat different.

```{r}
infoMat$TYPE
```

Most of the positions will be annotated as `snp`, however we have to allow for the possibility that more than one type of variant is observed at a particular location. 

```{r}
numVars <- unlist(lapply(infoMat$TYPE,length))
table(numVars)
```


We can find a few records where more than one type of variant is found.

```{r eval=FALSE}
infoMat[numVars>1,]
```

```{r echo=FALSE}
infoMat[numVars>1,][1:3,1:5]
```


## Accessing genotypes

The called genotypes can be accessed using the `geno` function. Recall that in the `.vcf` file, we have one column of genotype information for each sample, with each column consisting of `key:value` pairs. Using `geno` we can access all the values for a particular keys and be able to compare across samples. Usually each entry is `0/0` for a homozygous reference, `0/1` for a heterozygous call and `1/1` for a homozyous alternate allele.

```{r}
geno(hapmap.vcf)
head(geno(hapmap.vcf)$GT)
```

An entry of `.` indicates a position where no call could be made due to insufficient data. Moreover, we can also find `0/2` and `1/2` in rare cases where a second alternative allele was found. With `table` we can tabulate the calls between one sample and another

```{r}
table(geno(hapmap.vcf)$GT[,1])
table(geno(hapmap.vcf)$GT[,2])
table(geno(hapmap.vcf)$GT[,1], geno(hapmap.vcf)$GT[,2])
```

A common way to visualise these overlaps is to use a *venn diagram*. We can generate the venn diagram using the `limma` package (a popular package for the analysis of microarray data). First of all we need to create a data frame containing `0`s or `1` (or equivalenly `TRUE` or `FALSE`) depending on whether a variant was called in each sample. We can do this by creating three separate vectors and combining into a data frame. 

```{r}
called.NA12878 <- geno(hapmap.vcf)$GT[,1] != "0/0"
called.NA12873 <- geno(hapmap.vcf)$GT[,2] != "0/0"
called.NA12874 <- geno(hapmap.vcf)$GT[,3] != "0/0"

called <- data.frame(called.NA12878, called.NA12873,called.NA12874)
head(called)
head(geno(hapmap.vcf)$GT)
```

*N.B* a more sophisticated method would to use an `apply` statement

```{r}
called2 <- apply(geno(hapmap.vcf)$GT, 2,function(x) x !="0/0")
```

The `vennDiagram` function can now be used on the data frame we just created.

```{r}
limma::vennDiagram(called)
```


```{r}
all.hom <- which(geno(hapmap.vcf)$GT[,1]=="1/1" & geno(hapmap.vcf)$GT[,2]=="1/1" & geno(hapmap.vcf)$GT[,3]=="1/1")
geno(hapmap.vcf)$GT[all.hom,]
hapmap.vcf.homs <- hapmap.vcf[all.hom]
```


```{r}
writeVcf(hapmap.vcf.homs, filename = "hapmap.common.hom.vcf")
```


******

### Exercise

Find all the positions that were called as heterozyous in all samples
  - Write a vcf file

```{r echo=FALSE}

```


******

## Accessing ranges

The `rowRanges` function will retrieve the positions of our variants as a familiar `GRanges` object.

```{r}
rowRanges(hapmap.vcf)
```



```{r}
rowRanges(hapmap.vcf)[rowSums(called)==3]
```





```{r echo=FALSE}
table(isSNV(freebayes.calls)) 
table(isSNV(platypus.calls))
table(isSNV(freebayes.calls))/length(freebayes.calls)
table(isSNV(platypus.calls))/length(platypus.calls)

```

We can now restrict the two call sets to just SNVs
```{r}
freebayes.snv <- freebayes.calls[isSNV(freebayes.calls)]
length(freebayes.snv)
platypus.snv <- platypus.calls[isSNV(platypus.calls)]
length(platypus.snv)
```

******




## Genotype information

```{r}
freebayes.geno <- geno(freebayes.snv)
freebayes.geno
dp <- freebayes.geno$DP
ao <- freebayes.geno$AO
ro <- freebayes.geno$RO
head(dp)
head(ao)
head(ro)
plot(unlist(ao[,1]), unlist(ro[,1]))
```




The coordinates of the variants are given by `rowRanges`. The output is a familar `GRanges` object. Extra metadata included in this object include the reference (`REF`), alternate (`ALT`) alleles an a quality score. The quality score will depend on the caller being used.

```{r eval=FALSE}
rowRanges(freebayes.calls)
```

```{r echo=FALSE}
rowRanges(freebayes.calls)[1:3,]
```

We can use the `$` operator to extract particular properties of each variant; as if we were extracting the columns of a data frame.

```{r}
quals <- rowRanges(freebayes.calls)$QUAL
hist(quals)
```

The first stage of our analysis is to compute the overlap we have between our two callers (freebayes and platypus). To make our task a bit easier, we will consider just SNPs.

For instance, `freebayes` has called a mixture of SNVs and other events such as indels. Variants that are not SNVs are easy to identify by-eye because the `REF` or `ALT` allele does not have a length of 1.

```{r echo=FALSE}
rowRanges(freebayes.calls)[!isSNV(freebayes.calls)][1:3,]
```

Fortunately, `VariantAnnotation` has a useful function `isSNV` that we allow us to identify SNVs

```{r}
subset <- freebayes.calls[1:5]
isSNV(subset)
rowRanges(subset)
```




### Exercise

- Obtain the coordinates of the platypus and freebayes SNVs
- How many SNPs are in-common?
    + Recall the %over% function from yesterday....
- How many SNPS are unique to freebayes
    + unique to platypus?
- Does the quality score differ between calls between made by both tools, and those exclusive to freebayes?

******

```{r echo=FALSE}
rr.freebayes.snv <- rowRanges(freebayes.snv)
rr.platypus.snv <- rowRanges(platypus.snv)

inboth <- rr.freebayes.snv %over% rr.platypus.snv
sum(inboth)
sum(inboth) / length(freebayes.snv)

inboth2 <- rr.platypus.snv %over% rr.freebayes.snv
sum(inboth2)
sum(inboth2) / length(platypus.snv)

freebayes.only <- !rr.freebayes.snv %over% rr.platypus.snv
sum(freebayes.only)

platypus.only <- !rr.platypus.snv %over% rr.freebayes.snv
sum(platypus.only)

boxplot(rr.freebayes.snv$QUAL ~ inboth)

```

We might also expect that the depth observed at each variant is indicative of whether it will be called by both tools. We can test this by extracting the `INFO` from the `.vcf` file. Recall that this part of the file contains metadata that is *variant-specific*. 

```{r eval=FALSE}
info(freebayes.snv)
```

```{r echo=FALSE}
info(freebayes.snv)[,1:4]
```


```{r}
boxplot(info(freebayes.snv)$DP ~ inboth)
```

Finally, we define our set of common SNVs

```{r}
common.snv <- freebayes.snv[rr.freebayes.snv %over% rr.platypus.snv]
```


`VariantAnnotation` also has methods for being able to write an object to a `.vcf` file

```{r}
writeVcf(common.snv, file="common.snv.vcf")
```


## Locating variants

With our variants in `GRanges` form, it is possible to interact with the packages we saw yesterday. In particular, we can ask which of our variants are located within coding regions. As we described, transcript database packages are available from some organisms.

In order to make-use of this functionality, we first have to make sure that the names of the chromosomes are compatible with the database.

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
seqlevelsStyle(txdb)
seqlevelsStyle(common.snv) <- "UCSC"
common.snv <- keepSeqlevels(common.snv, "chr20")
```

With a bit of thought, it would be possible to perform overlaps with the exon structure described in the database. However, there is a convenience function `locateVariants` that will do the hard work for us. `CodingVariants` tells the function to look for variants within coding sequences. You can see the help for `locateVariants` (`?locateVariants`) for information on the various options.


```{r warning=FALSE}
locs <- locateVariants(common.snv, txdb, CodingVariants())
```

```{r eval=FALSE}
locs
```

```{r echo=FALSE}
locs[1:3]
```

If we want, we can locate all the variants with respect to the nearest feature (in the case of intergenic variants). 

```{r warning=FALSE, message=FALSE}
all.locs <- locateVariants(common.snv, txdb, AllVariants())
```

```{r eval=FALSE}
all.locs
table(all.locs$LOCATION)
```

```{r echo=FALSE}
all.locs[1:4]
table(all.locs$LOCATION)
```


******

### Exercise

Annotate the coding variants using the `org.Hs.eg.db` package that we saw yesterday

  - Add the gene symbol to give the following output

```{r echo=FALSE,message=FALSE}
library(org.Hs.eg.db)
symb <- select(org.Hs.eg.db, keys = locs$GENEID, keytype = "ENTREZID",columns="SYMBOL")
locs$SYMBOL <- symb[,2]
locs[1:3]
```

******

You may notice that the variants we have observed may affect more than one transcript, so can be listed more than once in the `locs` object. Also, in `locs` we no longer have information about the variant itself (for example the change from reference to alternative base that is observed). 

We might want to subset our set of variants to contain those found in coding regions. We can do this using the `QUERYID` variable, which is an index to the original vcf object. However, we need to take into account the fact that some variants are repeated.

```{r}
common.coding.snv <- common.snv[unique(locs$QUERYID)]
writeVcf(common.coding.snv, file="common.coding.snv.vcf")
common.coding.ranges <- rowRanges(common.coding.snv)
```


Given that we have found variants within coding sequences, we probably want to know whether these variants will have any impact on the amino acids produced. 

## Predicting Amino acid changes

Determining whether a change at a particular base will lead to a different amino acid can be performed using one of the many `BSgenome...` packages in Bioconductor. These packages provide a pre-built database of genome sequence for various organisms. They allow us to query particular regions of the genome and return the results in a `Biostrings`-compatible format. We have already seen that `Biostrings` allows us to perform various computations on DNA sequences; including the ability to translate DNA to RNA. 

There is a function, `predictCoding` that will do all the predictions for us. In order to make use of this function, we need to have loaded a pre-built genome in `Biostrings` format. There are many such packages listed on the bioconductor web page. The one we need is `BSgenome.Hsapiens.UCSC.hg19`

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
coding.pred <- predictCoding(common.coding.snv, txdb, seqSource=Hsapiens)
```

```{r eval=FALSE}
coding.pred 
table(coding.pred $CONSEQUENCE)
```

```{r echo=FALSE}
coding.pred [1:3]
table(coding.pred $CONSEQUENCE)
```


```{r}
coding.pred[coding.pred$CONSEQUENCE == "nonsense"]
```



******

### Exercise

Which exons have more than 2 coding variants?

- verify your answer in IGV

******

```{r echo=FALSE,message=FALSE}

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
exons <- exons(txdb)
exons.20 <- keepSeqlevels(exons,"chr20")
perExonSnp <- countOverlaps(exons.20, common.coding.ranges )
multiSnpExon <- exons.20[perExonSnp > 2]
multiSnpExon$ENTREZ <- select(txdb, keys=as.character(multiSnpExon$exon_id), keytype = "EXONID", columns="GENEID")[,2]
multiSnpExon$SYMBOL <- select(org.Hs.eg.db, keys=multiSnpExon$ENTREZ, keytype="ENTREZID", columns="SYMBOL")[,2]
multiSnpExon
```





# <a name="appendix"></a> Appendix

## Files used in session

Commands used to generate bam files for genotype calling

- First, get a copy of the reference genome from the 1000 genomes FTP site

```{}
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz -P ../ref_data/
gunzip ../ref_data/human_g1k_v37.fasta.gz

```

- We have already seen how `samtools` can be used to view aligned reads in `.bam` format. In fact, the `.bam` does not need to be stored locally on our desktop machine, and can be an FTP link. Here we `view` the remote link to the bam file and extract only the chromosome 22 reads. 

```{}
samtools view -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/other_exome_alignments/NA19239/exome_alignment/NA19239.mapped.solid.mosaik.YRI.exome.20111114.bam 22 | samtools view -bS - > data/hapmap/NA19239.chr22.bam
samtools index data/hapmap/NA19239.chr22.bam

```

```{}
samtools view -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/other_exome_alignments/NA19238/exome_alignment/NA19238.mapped.illumina.mosaik.YRI.exome.20111114.bam 22 | samtools view -bS - > data/hapmap/NA19238.chr22.bam
samtools index data/hapmap/NA19238.chr22.bam

```

```{}
samtools view -h ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/other_exome_alignments/NA19240/exome_alignment/NA19240.mapped.solid.mosaik.YRI.exome.20111114.bam 22 | samtools view -bS - > data/hapmap/NA19240.chr22.bam
samtools index data/hapmap/NA19240.chr22.bam
```


## Using the platypus caller

When learning about SNP-calling, it is worth being able to use more than one caller. In fact, it may be beneficial to use multiple-callers and take the consensus of the calls afterwards.

The [platypus](http://www.well.ox.ac.uk/platypus) caller from Oxford is another highly-regarded SNP-caller. For the purposes of this tutorial, is it also relatively easy to run

The command to run `platypus` with the default parameters is as follows:-

```{}
### N.B tidy-up the path to platypus
python /home/dunnin01/software/Platypus_0.8.1/Platypus.py callVariants --refFile ../ref_data/human_g1k_v37.fasta --bamFiles data/hapmap/NA12878.chr20.bam --regions 20 -o NA12878.chr20.platypus.vcf 

bgzip -c NA12878.chr20.platypus.vcf  > NA12878.chr20.platypus.vcf.gz
tabix -p  vcf NA12878.chr20.platypus.vcf.gz

```


As before, we can examine the contents of the `.vcf` file produced.

```{r, echo=FALSE,comment=NA}
cat(system("head NA12878.chr20.platypus.vcf",intern=TRUE),sep="\n")
```


```{r echo=FALSE,comment=NA}
cat(system("grep -v '##' NA12878.chr20.platypus.vcf | head -n5",intern = TRUE),sep = "\n")
```



## <a name=bedtools></a> Intersecting calls: using bedtools

[bedtools](http://bedtools.readthedocs.io/en/latest/) has some useful tools for overlapping various types of genomic intervals

We will use bedtools to refine our calls to a consensus set (i.e agreed by both callers) and present in exonic regions

```{}
bedtools intersect -header -a NA12878.chr20.freebayes.vcf -b NA12878.chr20.platypus.vcf > NA12878.chr20.consensus.vcf
```

******

### Exercise

- The -v option to intersect selects variants that are present in one file but not the other
- Create a vcf file contains calls made by freebayes, but not playtpus
- Create a vcf file contains calls made by platypus, but not freebayes


******


```{r echo=FALSE}
if(!file.exists("exons.chr20.bed")){
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(rtracklayer)
  library(org.Hs.eg.db)
  
  tx <- TxDb.Hsapiens.UCSC.hg19.knownGene
  exons <- exonsBy(tx, "gene")
  exons.20 <- keepSeqlevels(exons, "chr20")
  exons.20 <- unlist(exons.20)
  names(exons.20) <- select(org.Hs.eg.db, keys=names(exons.20), keytype = "ENTREZID",columns="SYMBOL")[,2]
  seqlevelsStyle(exons.20) <- "Ensembl"
  export(exons.20,con="exons.chr20.bed")
}
```

Selecting exonic calls only.....

```{}
bedtools intersect -header -a NA12878.chr20.consensus.vcf -b exons.chr20.bed > NA12878.chr20.exonic.consensus.vcf
```



# References

- [freebayes tutorial](http://clavius.bc.edu/~erik/CSHL-advanced-sequencing/freebayes-tutorial.html)
- [freebayes tutorial 2](https://libraries.io/github/benjaypunto/freebayes)
- [vcf files tutorial](https://faculty.washington.edu/browning/beagle/intro-to-vcf.html)
- [bedtools tutorial](http://quinlanlab.org/tutorials/bedtools/bedtools.html)